### ISSUE TITLE
Questionable call to ReadClipper.hardClipAdaptorSequence() in AssemblyBasedCallerUtils

### ISSUE THREAD

**Issue by droazen** [broadinstitute/gatk#3184]
@sooheelee has some serious concerns about `ReadClipper.hardClipAdaptorSequence()`, which is called in `Mutect2` and `HaplotypeCaller` via `AssemblyBasedCallerUtils.finalizeRegion()`. She thinks that the method being used to find the adaptor boundary for clipping purposes is completely bogus!

This is some pretty old code that was also in the GATK3 versions of these tools, so if it's crazy, then we can at least take "comfort" in the fact that it's been like this for a very long time...

**Comment by droazen** [comment]
@davidbenjamin Would you mind having a look at this when you have time?

**Comment by davidbenjamin** [comment]
@droazen @sooheelee Anything involving adaptors is not my forte but here goes my best shot.

The basic idea of `getAdaptorBoundary` is to find the end of the insert, that is, where the original fragment ends and the adaptor begins.  Since the 5' ends of the paired reads define the bounds of a fragment, its approach is to look for the start of the forward strand mate if our read is on the reverse strand, or the end of a reverse strand mate if our read is on the forward strand.  This protects us from the possibility of a read longer than its insert.  So far, so good, I think.

One possible source of error is that these boundaries are determined by the *alignment* starts of the paired reads.  This might clip too much if a read has a soft-clipped end and it turns out that these soft-clipped bases were e.g. a real insertion.  Then the soft-clipped bases would be considered part of the adaptor.  However, this can't do the more harmful thing of failing to clip an adaptor as far as I can tell.  So it looks to me like the only issue is the edge case of falsely soft-clipped bases in very short inserts, in which case we hard clip the soft clips and in theory could lose some sensitivity. 

That said, @sooheelee may have something else in mind, and @yfarjoun will have a better-informed opinion.

**Comment by yfarjoun** [comment]
I agree with @sooheelee here, it would be better to use the unclipped 5' end as in mark-duplicates.

also it would be better if the test
```
@Test(dataProvider = "HasWellDefinedFragmentSizeData")
    private void testHasWellDefinedFragmentSize(final String name, final GATKSAMRecord read, final boolean expected) {
        Assert.assertEquals(ReadUtils.hasWellDefinedFragmentSize(read), expected);
    }
```
were public, so that it would be active...

**Comment by davidbenjamin** [comment]
@yfarjoun I'm not understanding...  If we're on the reverse strand, then we reach the adaptor at the 5' end of the forward strand i.e. at one `getMateStart() + 1`, which is what the code does now.  If we're on the forward strand the equivalent logic would be `getMateEnd() + 1`, but no such method exists, so we use `read.getStart() + abs(read.getFragmentLength())`.  Why is this not equivalent?

**Comment by yfarjoun** [comment]
getFragmentLength is not well defined (unfortinately) there are two definition that are being used...one in the sam-spec, and one in Heng's code....

if the reads include a MC tag, I'd use that and call htsjdk htsjdk.samtools.SAMUtils#getMateAlignmentEnd 

if not, I guess we can use fragment length, but I'd put out a warning that it woudl work better if MC was present.

**Comment by davidbenjamin** [comment]
@yfarjoun Just to be clear, are you saying that `getMateAlignmentEnd() + 1` is ideal for forward strand reads but `read.getStart() + abs(read.getFragmentLength())` will have to do if the `MC` tag is missing, and that we can leave it as `getMateStart() + 1` for reverse strand reads?

@droazen A priori I would expect the additional cost of parsing each read's mate CIGAR to be negligible compared to other stuff we do but I also understand the virtue of being careful.  Given, however, that this is invoked for every assembled read in HaplotypeCaller it should suffice just to measure the wall clock time of HaplotypeCaller.  Would seeing no change there be enough?

**Comment by droazen** [comment]
@davidbenjamin I feel like we need to put good, continuous *performance* regression tests in place for the `HaplotypeCaller` so that we can make changes of this nature without fear. Testing for a performance regression in the `HaplotypeCaller` is currently very non-trivial -- you have to run with and without intervals, with and without -ERC GVCF, on both exome and genome to be confident that you haven't killed performance in a certain mode.

**Comment by lbergelson** [comment]
If this is blocking us from making otherwise useful changes then we should prioritize adding these tests...

**Comment by yfarjoun** [comment]
sorry, as with markduplicates, I'd use 

```
rec.getReadNegativeStrandFlag() ? rec.getUnclippedEnd() : rec.getUnclippedStart();
```
for the record itself, 
and 
```
record.getMateNegativeStrandFlag() ? SAMUtils.getMateUnclippedEnd(record) : SAMUtils.getMateUnclippedStart(record);
```
for the mate.

**Comment by meganshand** [comment]
@davidbenjamin This seems to now be a problem for `PalindromeArtifactClipReadTransformer` as well. In cases where there are soft clips you can miscalculate the `adaptorBoundary`. For mitochondria this causes edge case problems if you're on the end of the contig. I suppose for that specific case it could be fixed in `PalindromeArtifactClipReadTransformer` directly, but it makes more sense  to happen in `getAdaptorBoundary` (if cost was no issue).
